File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/harness-check-rule.ts
#!/usr/bin/env bun
// -*- coding: utf-8 -*-
import type {
  HarnessManifest,
  Manifest,
} from "../core/manifest";
import type { RuleContext } from "../core/rule-context";
import type { Severity } from "../core/severity";

/**
 * Strategy interface implemented by each harness check rule.
 */
export interface HarnessCheckRule {
  /**
   * Manifest category key used by this rule.
   */
  readonly category: string;

  /**
   * Determines whether this rule applies to the current context.
   */
  applies(ctx: RuleContext): boolean;

  /**
   * Validates the project against this rule.
   * Returns a read-only array of findings.
   */
  validate(ctx: RuleContext): readonly Finding[];

  /**
   * Auto-formats the project against this rule, when supported.
   * Returns the absolute paths of files modified by this rule.
   * Rules without an automatic fix MUST return an empty array.
   */
  format?(ctx: RuleContext): readonly string[];
}

/**
 * Describes whether a fix is safe to apply automatically or requires review.
 *
 * - `safe`: The fix can be applied automatically without user review.
 * - `unsafe`: The fix has potential side effects and requires manual review before application.
 * - `manual`: The fix cannot be automated and requires manual implementation by the user.
 */
export type FixSafety = "safe" | "unsafe" | "manual";

/**
 * A single edit operation within a fix, specifying the exact text replacement.
 */
export interface FindingEdit {
  /**
   * Absolute file path or relative path to the file containing the fix location.
   */
  readonly file: string;

  /**
   * Starting line number of the text to replace (1-indexed).
   */
  readonly startLine: number;

  /**
   * Starting column number of the text to replace (1-indexed).
   */
  readonly startColumn: number;

  /**
   * Ending line number of the text to replace (1-indexed).
   */
  readonly endLine: number;

  /**
   * Ending column number of the text to replace (1-indexed).
   */
  readonly endColumn: number;

  /**
   * The replacement text to insert at the specified location.
   */
  readonly replacement: string;
}

/**
 * Describes an automated fix for a validation finding.
 */
export interface FindingFix {
  /**
   * Human-readable description of what the fix does.
   */
  readonly description: string;

  /**
   * Categorization of whether the fix is safe to apply automatically.
   */
  readonly safety: FixSafety;

  /**
   * Optional array of edit operations that implement this fix.
   * When present, provides exact text replacements; when absent, the fix description alone
   * should guide manual resolution. Recommended when the rule can construct precise edits.
   */
  readonly edits?: readonly FindingEdit[];
}

/**
 * A validation finding emitted by a rule.
 */
export interface Finding {
  severity: Severity;
  category: string;
  message: string;

  /**
   * Absolute file path or relative path where the finding occurs (1-indexed locations).
   */
  readonly file?: string;

  /**
   * Starting line number of the finding (1-indexed).
   */
  readonly startLine?: number;

  /**
   * Starting column number of the finding (1-indexed).
   */
  readonly startColumn?: number;

  /**
   * Ending line number of the finding (1-indexed).
   */
  readonly endLine?: number;

  /**
   * Ending column number of the finding (1-indexed).
   */
  readonly endColumn?: number;

  /**
   * Optional fix information associated with this finding.
   */
  readonly fix?: FindingFix;
}

export type { HarnessManifest, Manifest, RuleContext, Severity };

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/harness_check_rule.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""
ABC for harness check rules.
"""

from __future__ import annotations

import sys


from abc import ABC, abstractmethod
from collections.abc import Iterable
from enum import Enum
from pathlib import Path
from typing import NamedTuple

import libcst as cst

from core.rule_context import RuleContext

sys.stdout.reconfigure(encoding="utf-8")
sys.stderr.reconfigure(encoding="utf-8")


class FixSafety(str, Enum):
    """
    Indicates how safely an auto-fix can be applied.

    :cvar SAFE: Fix preserves program semantics (e.g., simple formatting).
    :cvar UNSAFE: Fix may alter program behavior; requires review.
    :cvar MANUAL: No automatic fix is available; manual intervention required.
    """

    SAFE = "safe"
    UNSAFE = "unsafe"
    MANUAL = "manual"


class FindingEdit(NamedTuple):
    """
    A concrete text edit that a fix would apply.

    :ivar file: Repository-relative POSIX path of the file to edit.
    :ivar start_line: 1-indexed start line.
    :ivar start_column: 1-indexed start column.
    :ivar end_line: 1-indexed end line (inclusive).
    :ivar end_column: 1-indexed end column (exclusive).
    :ivar replacement: Text that replaces the selected range.
    """

    file: str
    start_line: int
    start_column: int
    end_line: int
    end_column: int
    replacement: str


class FindingFix(NamedTuple):
    """
    Describes how a rule violation would be fixed.

    :ivar description: One-line human-readable description of the fix.
    :ivar safety: Safety classification of the fix.
    :ivar edits: Optional list of concrete edits; empty when the rule cannot
        provide exact text spans.
    """

    description: str
    safety: FixSafety
    edits: tuple[FindingEdit, ...] = ()


class Finding(NamedTuple):
    """
    A validation finding emitted by a rule.

    :ivar severity: Severity level (ERROR/WARN/INFO).
    :ivar category: Manifest category key.
    :ivar message: Human-readable violation message.
    :ivar file: Repository-relative POSIX path of the offending file (optional).
    :ivar start_line: 1-indexed start line (optional).
    :ivar start_column: 1-indexed start column (optional).
    :ivar end_line: 1-indexed end line (optional).
    :ivar end_column: 1-indexed end column (optional).
    :ivar fix: Fix metadata, when an automatic or assistive fix is known.
    """

    severity: str
    category: str
    message: str
    file: str | None = None
    start_line: int | None = None
    start_column: int | None = None
    end_line: int | None = None
    end_column: int | None = None
    fix: FindingFix | None = None


class HarnessCheckRule(ABC):
    """Strategy ABC implemented by each validation rule."""

    @abstractmethod
    def applies(self, ctx: RuleContext) -> bool:
        """
        Check if this rule applies to the context.

        :param ctx: rule execution context.
        :returns: ``True`` when the rule should run against ``ctx``.
        """

    @abstractmethod
    def validate(self, ctx: RuleContext) -> Iterable[Finding]:
        """
        Validate the project and return findings.

        :param ctx: rule execution context.
        :returns: iterable of findings; empty when no issues are found.
        """

    def format(self, ctx: RuleContext) -> Iterable[Path]:
        """
        Auto-format the project against this rule, when supported.

        Rules without an automatic fix MUST keep this default and return an
        empty iterable.

        :param ctx: rule execution context.
        :returns: iterable of absolute paths modified by this rule.
        """
        return ()


def parse_python(path: Path) -> tuple[cst.Module | None, str | None]:
    """Parse a Python file and return Module or error message."""
    if not path.is_file():
        return (None, "file not found")
    module = cst.parse_module(path.read_text(encoding="utf-8"))
    return (module, None)


def has_nested_function(func_node: cst.FunctionDef) -> bool:
    """Check if function body contains nested function, class, or lambda."""

    class NestedFinder(cst.CSTVisitor):
        def __init__(self) -> None:
            super().__init__()
            self.found = False
            self.depth = 0

        def visit_FunctionDef(self, node: cst.FunctionDef) -> bool:
            if self.depth > 0:
                self.found = True
                return False
            self.depth += 1
            return True

        def visit_ClassDef(self, node: cst.ClassDef) -> bool:
            if self.depth > 0:
                self.found = True
                return False
            return True

        def leave_FunctionDef(self, original: cst.FunctionDef) -> None:
            self.depth -= 1

        def visit_Lambda(self, node: cst.Lambda) -> bool:
            self.found = True
            return False

    visitor = NestedFinder()
    func_node.visit(visitor)
    return visitor.found
